### PR TITLE
Enable min size 0 in AWS

### LINKED ISSUE #132: Allow scaling to/from 0 on AWS
Follow up of #43 on AWS.

cc: @mumoshu @7chenko

### PR DESCRIPTION
**Pull request by sethpollack**
fixes #132 part of #43

### CODE REVIEW

**Comment by mwielgus** [comment]
`
Skipping hack/../hack/verify-all.sh
!!! 'gofmt -s' needs to be run on the following files: 
./cluster-autoscaler/cloudprovider/aws/ec2_instance_types.go
!!! 'golint' problems: 
/home/travis/gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/utils.go:31:2: exported const DefaultArch should have comment (or a comment on this block) or be unexported
/home/travis/gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/utils.go:37:1: exported function BuildReadyConditions should have comment or be unexported
/home/travis/gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/utils.go:68:1: comment on exported function BuildKubeProxy should be of the form "BuildKubeProxy ..."
/home/travis/gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/utils.go:111:1: exported function JoinStringMaps should have comment or be unexported /home/travis/gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/ec2_instance_types.go:27:5: exported var InstanceTypes should have comment or be unexported
`

**Comment by mwielgus** [comment]
cc: @ryanwalls @mumoshu

**Comment by mumoshu** [comment]
@sethpollack You can run `make format` in the project root for resolving the `'gofmt -s' needs to be run...` errors 😉

File: cluster-autoscaler/cloudprovider/aws/aws_manager.go [hunk 1]
```diff
@@ -163,3 +176,82 @@ func (m *AwsManager) GetAsgNodes(asg *Asg) ([]string, error) {
...
+		SelfLink: fmt.Sprintf("/api/v1/nodes/%s", nodeName),
+		Labels:   map[string]string{},
+	}
+
+	node.Status = apiv1.NodeStatus{
+		Capacity: apiv1.ResourceList{},
+	}
+
+	// TODO: get a real value.
+	node.Status.Capacity[apiv1.ResourcePods] = *resource.NewQuantity(110, resource.DecimalSI)
```
Comment (mumoshu) [inline cluster-autoscaler/cloudprovider/aws/aws_manager.go]: @mwielgus Just wondering but would you mind sharing me how we can calculate the capacity for ResourcePods? Is there any formula or api for that?

File: cluster-autoscaler/cloudprovider/aws/aws_manager.go [hunk 2]
```diff
@@ -163,3 +176,82 @@ func (m *AwsManager) GetAsgNodes(asg *Asg) ([]string, error) {
...
+
+	node.Status.Conditions = cloudprovider.BuildReadyConditions()
+	return &node, nil
+}
+
+func buildGenericLabels(template *asgTemplate, nodeName string) map[string]string {
+	result := make(map[string]string)
+	// TODO: extract it somehow
+	result[kubeletapis.LabelArch] = cloudprovider.DefaultArch
+	result[kubeletapis.LabelOS] = cloudprovider.DefaultOS
```
Comment (mumoshu) [inline cluster-autoscaler/cloudprovider/aws/aws_manager.go]: @mwielgus Would you mind sharing me when we need "correct" values for these labels in practice? Only when the user wants to utilize these in nodeSelector in a pod spec?

**Comment by sethpollack** [comment]
@mumoshu Thanks! It's a generated file, so its probably better if I fix the template. I'll push the fixes soon.

**Comment by mumoshu** [comment]
FYI: pushed `mumoshu/cluster-autoscaler:aws_min_zero_1` to Docker Hub for testing

File: cluster-autoscaler/cloudprovider/aws/ec2_instance_types/gen.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	"{{ .InstanceType }}": {
+		InstanceType: "{{ .InstanceType }}",
+		VCPU:         {{ .VCPU }},
+		MemoryMb:     {{ .Memory }},
+	},
+{{- end }}
+}
+`))
+
+func main() {
```
Comment (mumoshu) [inline cluster-autoscaler/cloudprovider/aws/ec2_instance_types/gen.go]: As we're using glog for logging in this, don't we need `flag.Parse()` in the very beginning of the `main` func for glog to be configured via command-line flags?

File: cluster-autoscaler/cloudprovider/aws/ec2_instance_types/gen.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	instanceTypes := make(map[string]*instanceType)
```
Comment (mumoshu) [inline cluster-autoscaler/cloudprovider/aws/ec2_instance_types/gen.go]: Sorry If I'm not following you correctly but for me, it seems like we are overwriting instance types over and over again for every region into this map?

Perhaps we should make the keys of the `instanceTypes` map region-aware, or we end up with throwing away almost all the instance type details provided but for the last region(which was `us-gov-west-1` for me in the example below):

```
$ go run ec2_instance_types/gen.go -logtostderr
I0705 17:15:01.732876   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/ap-northeast-1/index.json
I0705 17:15:02.711280   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/us-west-2/index.json
I0705 17:15:03.444548   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/eu-west-2/index.json
I0705 17:15:03.820958   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/ap-south-1/index.json
I0705 17:15:04.286667   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/us-west-1/index.json
I0705 17:15:04.889539   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/ap-southeast-1/index.json
I0705 17:15:05.548524   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/eu-central-1/index.json
I0705 17:15:06.151498   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/ca-central-1/index.json
I0705 17:15:06.532968   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/us-east-1/index.json
I0705 17:15:07.211180   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/ap-northeast-2/index.json
I0705 17:15:07.710261   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/us-east-2/index.json
I0705 17:15:08.199422   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/eu-west-1/index.json
I0705 17:15:08.892445   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/ap-southeast-2/index.json
I0705 17:15:09.559583   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/sa-east-1/index.json
I0705 17:15:10.138488   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/cn-north-1/index.json
I0705 17:15:10.152407   21476 gen.go:124] Error unmarshaling https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/cn-north-1/index.json skipping...
I0705 17:15:10.152443   21476 gen.go:106] fetching https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/us-gov-west-1/index.json
```

Comment (mumoshu) [inline cluster-autoscaler/cloudprovider/aws/ec2_instance_types/gen.go]: Ah, or is this an intended behavior, assuming all the instance types named the same to have equally sized cpu and memory across different regions?

**Comment by mumoshu** [comment]
@sethpollack I have just verified that CA built from your branch successfully scale the node number (up from|down to) zero on my test cluster. Great job 👍

File: cluster-autoscaler/cloudprovider/aws/aws_manager.go [hunk 5]
```diff
@@ -163,3 +176,82 @@ func (m *AwsManager) GetAsgNodes(asg *Asg) ([]string, error) {
...
+
+	result[kubeletapis.LabelInstanceType] = template.InstanceType.InstanceType
+
+	result[kubeletapis.LabelZoneRegion] = template.Region
+	result[kubeletapis.LabelHostname] = nodeName
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/aws/aws_manager.go]: Shouldn't this also populate LabelZoneFailureDomain?

File: cluster-autoscaler/cloudprovider/aws/ec2_instance_types/gen.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+	resolver := endpoints.DefaultResolver()
+	partitions := resolver.(endpoints.EnumPartitions).Partitions()
+
+	for _, p := range partitions {
+		for _, r := range p.Regions() {
+			url := "https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/" + r.ID() + "/index.json"
+			glog.Infof("fetching %s\n", url)
+			res, err := http.Get(url)
+			if err != nil {
+				glog.Infof("Error fetching %s skipping...\n", url)
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/aws/ec2_instance_types/gen.go]: Shouldn't that be at least a warning level log (same for other errors in this loop)?

File: cluster-autoscaler/cloudprovider/aws/ec2_instance_types/gen.go (hunk 4 above)
Comment (sethpollack) [inline cluster-autoscaler/cloudprovider/aws/ec2_instance_types/gen.go]: Correct

File: cluster-autoscaler/cloudprovider/aws/aws_manager.go (hunk 5 above)
Comment (sethpollack) [inline cluster-autoscaler/cloudprovider/aws/aws_manager.go]: The ASG can have multiple zones, should I just pick a random one?

File: cluster-autoscaler/cloudprovider/aws/ec2_instance_types/gen.go (hunk 6 above)
Comment (sethpollack) [inline cluster-autoscaler/cloudprovider/aws/ec2_instance_types/gen.go]: Ok, I'll fix

**Comment by sethpollack** [comment]
@mumoshu Nice thanks! How did you test it, did you use a `nodeSelector` and ASG tags?

File: cluster-autoscaler/cloudprovider/aws/aws_manager.go (hunk 5 above)
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/aws/aws_manager.go]: Good point. I think it should log some warning if there are multiple zones and just pick one of them. 

My rationale is as follows - we already do basically the same thing if there are any nodes in the ASG. We just pick a random node, copy it and believe a new one will be identical (except for name, etc). If there are nodes in multiple zones in the same ASG we effectively just pick a random zone. 

The supported way of having multi-zone cluster is to have a separate ASG in every zone and run CA with `--balance-similar-node-groups=true`. We should probably improve the documentation about that, but that's a separate issue.

WDYT @mwielgus, @mumoshu ?

Comment (mumoshu) [inline cluster-autoscaler/cloudprovider/aws/aws_manager.go]: @MaciekPytel @sethpollack I agree with your idea - we should pick a random one with emitting a BIG warning message as having an ASG spreading over multiple AZs is something we don't recommend in the first place.

**Comment by mumoshu** [comment]
@sethpollack No, `nodeSelector` was not necessary in my test setup. Also - just curious but how ASG tags relates here?

What I did is as follows.

I've created 5 varying node pools with kube-aws to verify that it works even in a needlessly complex setup 😉

3 of 5 are ASGs and 2 are spot fleets. Only one of 3 ASGs is recognized as CA's scaling target by enabling CA's node-group-auto-discovery feature on it.

With the setup, I've removed all the pods from the node group and then observed that CA eventually scales down the cluster by terminating the only node from the ASG i.e. its desired capacity is decreased to zero.

After that, I've created enough pods to ensure there are some pending pods(this is why `nodeSelector` was not necessary. No need to use that in order to make some pendings pods assigned to the node group on which CA is enabled on) and then observed that CA eventually scales up the cluster by adding nodes to the ASG i.e. its desired capacity is changed from zero to non-zero(actually to the max size of the ASG).

**Comment by sethpollack** [comment]
@mumoshu So I don't think that there is a way to pull node labels like with GCE, so I added the labels as tags to the ASG and I'm puling those in as labels. It's a bit hacky, but i'm not sure of a better way, any ideas?

**Comment by mumoshu** [comment]
@sethpollack Ah, I began to understand your concern now.

AFAIK, ASG tags are not automatically translated or copied into node labels, right?

Then, I guess we should provide the user an option to specify(opt-in) exactly which node labels should be included in the node template, via tags as you've suggested.

More concretely, I'd suggest:

```golang
func extractLabelsFromAsg(tags []*autoscaling.TagDescription) map[string]string {
	result := make(map[string]string)

	for _, tag := range tags {
                k := *tag.Key
                v := *tag.Value
                splits := strings.Split(k, "k8s.io/cluster-autoscaler/node-template/label/")
                if len(splits) > 1 {
                    label := splits[1]
                    if label != "" {
	                result[label] = v
                    }
                }
	}

	return result
}
```

As long as it is well documented, the user can specify which label should be included in the node template. A tag with the key `k8s.io/cluster-autoscaler/node-template/label/foo` and the value `bar` results in a label `foo=bar` in the node template. WDYT?

**Comment by mumoshu** [comment]
Also note that a k8s cluster provisioner like kops, kube-aws, etc can run a long-running daemon to periodically sync node labels as tags prefixed with `k8s.io/cluster-autoscaler/node-template/label/`.

**Comment by sethpollack** [comment]
@mumoshu I like that, I will make those changes.

**Comment by sethpollack** [comment]
@justinsb any thoughts on this?

**Comment by sharpevo** [comment]
@mumoshu How did you

**Comment by sharpevo** [comment]
@mumoshu How did you scale down the node number to zero?

After pulling images via `docker pull mumoshu/cluster-autoscaler:aws_min_zero_1`,  the cluster started a new deployment by a CA yaml file, with the commands as following
```
        command:
            - ./cluster-autoscaler
            ...
            - --nodes=0:3:asg-kube-nodes
            ...
```
Then CA complains that the MinSize must be greater than one:
```
Failed to create AWS cloud provider: failed to parse node group spec: invalid node group spec: min size must be >= 1
```
So I increase the MinSize to 1,  and finally, the configurations are
```
# Cluster Autoscaler
--nodes=1:3:asg-kube-nodes

# AWS ASG
"MinSize": 0,
"MaxSize": 5,
"DesiredCapacity": 0,
```
As a result, the ASG scale up nodes from zero to one successfully, but failed to scale down to zero:
```
# Scale up

Pod default/targetseq-20170706-093850-8l7l6 is unschedulable
Upcoming 0 nodes
Best option to resize: asg-kube-nodes
Estimated 1 nodes needed in asg-kube-nodes
Final scale-up plan: [{asg-kube-nodes 0->1 (max: 3)}]
Setting asg asg-kube-nodes size to 1
...

# Scale down
ip-172-31-23-142.cn-north-1.compute.internal was unneeded for 1m0.606076653s
Skipping ip-172-31-23-142.cn-north-1.compute.internal - node group min size reached
```

**Comment by mumoshu** [comment]
I used the auto discovery feature which can be enabled via
`--node-group-auto-discovery`.

Regarding the error log you've seen, probably we need to fix node group
spec parsing/validation logic not to reject the min size of zero?
2017年7月7日(金) 10:00 Yang Wu <notifications@github.com>:

**Comment by MaciekPytel** [comment]
Agreed, this should work with statically configured ASGs. I think this can be fixed by changing second parameter to true in https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go#L299.

@sethpollack can you take a look if that makes sense and include it in this PR?

**Comment by sethpollack** [comment]
@MaciekPytel I don't think that is the issue, it is already included in my PR.

**Comment by MaciekPytel** [comment]
@sethpollack Sorry, you're right of course. I just looked at log message and assumed that's the issue.

**Comment by mumoshu** [comment]
@sethpollack Sorry, so am I! The issue shouldn't happen then.

@sharpevo Just wondering but I don't understand the following sentence of yours:

> After pulling images via docker pull mumoshu/cluster-autoscaler:aws_min_zero_1

You don't need to manually `docker pull` the image to run CA?
Would you mind sharing a complete copy of the CA yaml file you've used for deploying CA?

**Comment by sharpevo** [comment]
@mumoshu 

I was trying to build images based on the [4a86058](https://github.com/sethpollack/autoscaler/commit/4a86058b93658ab30d830fc93183623d9655fec1) of this PR,  unfortunately it does not work well. So I decide to take a test with the image you built, and now everything's fine!

Here's the CA yaml file. Previously, when CA does not support scaling dowon to zero on AWS, I attach master node to the ASG with the `kube-system` namespace, in order to avoid being killed accidentally. Currently, it should not be necessary.

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: cluster-autoscaler
  namespace: kube-system
  labels:
    k8s-app: cluster-autoscaler
    version: 
spec:
  replicas: 1
  selector:
    matchLabels:
      k8s-app: cluster-autoscaler
      version: aws_min_zero
  template:
    metadata:
      labels:
        k8s-app: cluster-autoscaler
        version: aws_min_zero
    spec:
      containers:
        - image: reg.muffin.com/cluster/cluster-autoscaler:aws_min_zero
          name: cluster-autoscaler
          resources:
            limits:
              cpu: 100m
              memory: 300Mi
            requests:
              cpu: 100m
              memory: 300Mi
          command:
            - ./cluster-autoscaler
            - --v=4
            - --log-dir=/var/log
            - --cloud-provider=aws
            - --skip-nodes-with-local-storage=true
            #- --nodes=1:3:asg-kube-nodes
            - --node-group-auto-discovery=asg:tag=ca-discovery
            - --scale-down-delay=1m
            - --scale-down-unneeded-time=1m
            - --unregistered-node-removal-time=5m
            - --kubernetes=http://172.16.16.253:8080
          env:
            - name: AWS_REGION
              value: cn-north-1
          volumeMounts:
            - name: ssl-certs
              mountPath: /etc/ssl/certs/ca-certificates.crt
              readOnly: true
            - name: log
              mountPath: /var/log
              readOnly: false
            - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
              name: default-token
              readOnly: true
          imagePullPolicy: "Never"
      nodeSelector:
        nodeRole: master
      volumes:
        - name: ssl-certs
          hostPath:
            path: "/etc/ssl/certs/ca-bundle.crt"
        - name: log
          hostPath:
            path: "/var/log/cluster/cluster-autoscaler"
        - name: default-token
          secret:
            secretName: default-token-kn72h

```

**Comment by mumoshu** [comment]
@sharpevo Glad to hear that it worked for you 🎉

**Comment by MaciekPytel** [comment]
@sethpollack have you managed to run it successfully with static node group config? A comment with some description of how you tested it would be appreciated (as per our FAQ https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-should-i-test-my-code-before-submitting-pr).

The code and documentation looks good to me and if someone confirms it works with static node group config I'm ok to merge it.

**Comment by sethpollack** [comment]
Yes, I tested with
```command:
- ./cluster-autoscaler
- --cloud-provider=aws
- --nodes=0:1:nodes.example.com
```

**Comment by MaciekPytel** [comment]
/lgtm

@sethpollack can you squash it into one or two commits (1 implementation, 1 docs, or just one for everything)?

@ryanwalls @mumoshu @sharpevo @justinsb - do you have any further comments? If not I will merge it tomorrow.

**Comment by ryanwalls** [comment]
LGTM, thanks @sethpollack

**Comment by ryanwalls** [comment]
@sethpollack @mwielgus I'm relatively new to the k8s community.  What's the process for getting a 0.6.1 tag built and pushed to gcr.io/google_containers/cluster-autoscaler now that this has been merged?

**Comment by MaciekPytel** [comment]
@sethpollack - I've merged a PR by @aleksandra-malinowska adding taints to node template used for scale-from-0 in GCP/GKE: #172. Are you still interested in adding something similar for AWS?

**Comment by sethpollack** [comment]
Yeah, I can try to add that.

**Pull request merged**